parlai/core/metrics.py

```diff
@@ -42,6 +47,95 @@
 re_punc = re.compile(r'[!"#$%&()*+,-./:;<=>?@\[\]\\^`{|}~_\']')
 
 
+class Metric(ABC):
+    """
+    Base class for storing metrics.
+
+    Subclasses should define .value().
+    """
+
+    @abstractmethod
+    def value(self) -> float:
+        """
+        Return the value of the metric contained by the metric object, usually a scalar.
+        (For instance, if the metric object is SumMetric, .value() will return the sum 
+        stored by the object.)
+        """
+        pass
+
+    def __str__(self) -> str:
+        return f'{self.value():.4g}'
+
+    def __repr__(self) -> str:
+        return f'{self.__class__} ({self.value():.4g})'
+
+    @staticmethod
+    def as_number(obj: Any) -> Number:
```

Comment:
maybe we want as_float in addition to as_number?

Also obj should probably be explicitly a Tensor, a float, or an int.

It's shitty that mypy doesn't properly recognize float is a number. We might just want to use Union[float, int]. I do that elsewhere for the same reason (and we don't necessarily guarantee we can support complex numbers, for example)

Reply:
Yes, agreed - added as_float() and as_int() methods, and I'm now more explicit with obj's type